gk/main.c

```diff
@@ -1052,12 +1052,16 @@ xmit_icmp(struct gatekeeper_if *iface, struct ipacket *packet,
 		iface->l2_len_out);
 	/*
 	 * The IP header checksum filed must be set to 0
-	 * in order to offload the checksum calculation.
+	 * in order to calculate the checksum.
 	 */
 	icmp_ipv4->hdr_checksum = 0;
 	pkt->l2_len = iface->l2_len_out;
 	pkt->l3_len = sizeof(struct rte_ipv4_hdr);
-	pkt->ol_flags |= PKT_TX_IPV4 | PKT_TX_IP_CKSUM;
+	pkt->ol_flags |= PKT_TX_IPV4;
+	if (likely(iface->ipv4_hw_cksum))
+		pkt->ol_flags |=  PKT_TX_IP_CKSUM;
+	else
+		icmp_ipv4->hdr_checksum = rte_ipv4_cksum(icmp_ipv4);
```

Comment:
This code pattern is repeated several times. My suggestion is to add the following function and reuse it:
```c
inline void
set_ipv4_checksum(struct gatekeeper_if *iface, struct rte_mbuf *pkt,
	struct rte_ipv4_hdr *ipv4)
{
	ipv4->hdr_checksum = 0;
	pkt->ol_flags |= PKT_TX_IPV4;
	if (likely(iface->ipv4_hw_cksum))
		pkt->ol_flags |=  PKT_TX_IP_CKSUM;
	else
		ipv4->hdr_checksum = rte_ipv4_cksum(ipv4);
}
```